### ISSUE TITLE
Improve performance in astropy.coordinates

### ISSUE THREAD

**Issue by astrofrog** [astropy/astropy#1717]
This issue is to brainstorm on ideas for _significantly_ improving performance in astropy.coordinates (not for arrays, which are efficient, but for small number or single coordinates). This is motivated by the fact that for equivalent transformations, other packages can be 100 times (or more) faster. So this is not for improvements of 10-20% - we're talking one or more orders of magnitude improvements.

Initial ideas/examples:
- using `Quantity` internally can lead to significant slowdowns in places, for example:

```
In [12]: from astropy.coordinates import Angle

In [13]: a = Angle(3.3, unit='deg')

In [14]: %timeit np.sin(a)
10000 loops, best of 3: 129 us per loop

In [15]: %timeit np.sin(np.radians(3.3))
100000 loops, best of 3: 4.24 us per loop
```

So trigonometric functions applied to single quantities could be made 30 times faster if we were not using `Quantity` (or maybe `Quantity` could be made more performant).
- Transformations could be cached - that is, if one needs to do many transformations with the same parameters, but they need to be done sequentially (e.g. as @ejeschke needs), one could presumably compute the full transformation matrix the first time then cache it for future use.
- At the moment, I think that coordinates are converted back and forth between cartesian and spherical, but for some chains of transformations, it may be possible to first compute the different matrices (if they are dynamic) and then combine them, before applying them to the cartesian coordinates in one go. This also makes it possible to easily cache the matrix for a chain of transformations. So just to be clear, this means that we don't convert the coordinates from FK5 to FK4 then FK4 to Galactic - we find the transformation from FK5 to FK4, then FK4 to Galactic (for the appropriate equinoxes/epochs) then combine the matrices, and apply them in one go.
- Use ERFA C code where appropriate
- As a last resort, see if any of the code can be sped up with C or Cython.

**Comment by cdeil** [comment]
Related: http://stackoverflow.com/questions/36146183/astropy-skycoord-extremely-slow-how-to-resovle-it

Anyone wants to comment there?

**Comment by pllim** [comment]
Astropy on stackoverflow :scream: :clap:

**Comment by adrn** [comment]
I think many of these issues are now somewhat addressed in 3.1, so I think we should close this and open a new issue when we're ready to discuss new avenues for performance enhancements?
